File: server/controllers/systemConfigController.js
const systemConfigModel = require("../models/systemConfigModel");

/* ==============================
   GET all configurations
============================== */
exports.getAllConfigs = async (req, res) => {
  try {
    const configs = await systemConfigModel.getAllConfigs();
    res.json(configs);
  } catch (err) {
    console.error("❌ Error fetching configs:", err);
    res.status(500).json({ error: "Internal Server Error" });
  }
};

/* ==============================
   CREATE configuration
============================== */
exports.createConfig = async (req, res) => {
  try {
    const config = await systemConfigModel.createConfig(req.body);
    res.json(config);
  } catch (err) {
    if (err.code === "23505") {
      return res
        .status(400)
        .json({ error: `Academic year already exists.` });
    }
    console.error("❌ Error creating config:", err);
    res.status(500).json({ error: "Internal Server Error" });
  }
};

/* ==============================
   UPDATE configuration
============================== */
exports.editConfig = async (req, res) => {
  const { id } = req.params;

  if (!id || isNaN(Number(id))) {
    return res.status(400).json({ error: "Invalid or missing config ID" });
  }

  try {
    const updated = await systemConfigModel.updateConfig(id, req.body);

    if (!updated) {
      return res.status(404).json({ error: "Configuration not found" });
    }

    res.json(updated);
  } catch (err) {
    if (err.code === "23505") {
      return res
        .status(400)
        .json({ error: `Academic year already exists.` });
    }
    console.error("❌ Error updating config:", err);
    res.status(500).json({ error: "Internal Server Error" });
  }
};

/* ==============================
   DELETE configuration
============================== */
exports.deleteConfig = async (req, res) => {
  try {
    const deleted = await systemConfigModel.deleteConfig(req.params.id);

    if (!deleted) {
      return res.status(404).json({ error: "Configuration not found" });
    }

    res.json({ message: "Configuration deleted successfully" });
  } catch (err) {
    console.error("❌ Error deleting config:", err);
    res.status(500).json({ error: "Internal Server Error" });
  }
};

/* ==============================
   ACTIVATE configuration
============================== */
exports.activateConfig = async (req, res) => {
  try {
    const activated = await systemConfigModel.activateConfig(req.params.id);

    if (!activated) {
      return res.status(404).json({ error: "Configuration not found" });
    }

    res.json(activated);
  } catch (err) {
    console.error("❌ Error activating config:", err);
    res.status(500).json({ error: "Internal Server Error" });
  }
};

/* ==============================
   GET active configurations
============================== */
exports.getActiveConfigs = async (req, res) => {
  try {
    const configs = await systemConfigModel.getActiveConfigs();
    res.json(configs);
  } catch (err) {
    console.error("❌ Error fetching active configs:", err);
    res.status(500).json({ error: "Internal Server Error" });
  }
};

File: java-backend/imas-backend/src/main/java/com/rcf/imas/modules/masterdata/web/SystemConfigController.java
package com.rcf.imas.modules.masterdata.web;

import com.rcf.imas.modules.masterdata.persistence.SystemConfigRepository;
import com.rcf.imas.platform.error.ApiException;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/system-config")
// Node's system-config mount had NO auth. Mutations are ADMIN (below). `/active` stays role-agnostic
// (isAuthenticated) because the app-wide SystemConfigContext reads it for EVERY authenticated role on load
// -- gating it ADMIN would 403 the whole non-admin app. `/all` (full config list) is ADMIN (admin-only screen).
@PreAuthorize("isAuthenticated()")
class SystemConfigController {

    record ConfigRequest(String academicYear, String phase, Boolean isActive) {}
    // NOTE: global snake_case Jackson maps academic_year -> academicYear, is_active -> isActive
    // automatically — the frontend sends snake_case keys, so NO @JsonProperty pins here.

    private final SystemConfigRepository repo;

    SystemConfigController(SystemConfigRepository repo) { this.repo = repo; }

    @PreAuthorize("hasRole('ADMIN')")   // full config list is admin-only (Admin/SystemConfig.js)
    @GetMapping("/all")
    public List<Map<String, Object>> all() { return repo.findAll(); }

    @GetMapping("/active")
    public List<Map<String, Object>> active() { return repo.findActive(); }

    @PreAuthorize("hasRole('ADMIN')")
    @PostMapping
    public Map<String, Object> create(@RequestBody ConfigRequest req) {
        try {
            return repo.insert(req.academicYear(), req.phase(), req.isActive());
        } catch (DuplicateKeyException e) {
            throw ApiException.error(400, "Academic year already exists.");
        }
    }

    @PreAuthorize("hasRole('ADMIN')")
    @PutMapping("/{id}")
    public Map<String, Object> update(@PathVariable String id, @RequestBody ConfigRequest req) {
        long configId = parseId(id);
        try {
            return repo.update(configId, req.academicYear(), req.phase(), req.isActive())
                    .orElseThrow(() -> ApiException.error(404, "Configuration not found"));
        } catch (DuplicateKeyException e) {
            throw ApiException.error(400, "Academic year already exists.");
        }
    }

    @PreAuthorize("hasRole('ADMIN')")
    @DeleteMapping("/{id}")
    public Map<String, Object> delete(@PathVariable String id) {
        repo.delete(parseId(id)).orElseThrow(() -> ApiException.error(404, "Configuration not found"));
        return Map.of("message", "Configuration deleted successfully");
    }

    @PreAuthorize("hasRole('ADMIN')")
    @PutMapping("/{id}/activate")
    public Map<String, Object> activate(@PathVariable String id) {
        return repo.activate(parseId(id))
                .orElseThrow(() -> ApiException.error(404, "Configuration not found"));
    }

    private static long parseId(String id) {
        try {
            return Long.parseLong(id);
        } catch (NumberFormatException e) {
            throw ApiException.error(400, "Invalid or missing config ID");
        }
    }
}
